### DISCUSSION TITLE
Sampler orders are not consistent across backends

### DISCUSSION THREAD

**Discussion by kalomaze** [ggml-org/llama.cpp#3914] [General]
It was observed in my [Min P pull request](https://github.com/ggerganov/llama.cpp/pull/3841) that llama.cpp is doing something unique with the sampler order in comparison to other LLM inference backends:

<img width="532" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/66376113/4b79bda2-3dc3-4b37-a991-66b96c4fbbd7">

The koboldcpp fork has its own solution for this: an internal 'sampler order' that can be configured by the end user to accomodate for their own personal order, since the transformers library expects that the temperature always comes first.

<img width="275" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/66376113/8229fba0-ce65-4ba0-ac56-78f010849c11">

This allows the user to use the 'official' order where Temperature comes first as desired, or the alternative (that llama.cpp currently assumes) where Temp comes last.

Modifying this order has a **significant** impact in how the temp changes the model's outputs; if Temperature comes last with a 'safe' Min P or Top P, for example, it will still read correctly even with an obscenely high temperature value. But if Temperature comes _first_ as is expected in the Transformers library (as well as the official GPT2 implementation), this value will naturally have a _much_ different effect, as the actual values will have been modified before those truncation-based samplers were run, causing them to operate on a different 'scale'. I.e a token that would've been considered as under the 'Minimum' in a normal Min P setup might be over the required % if a high temperature was applied first.

The lack of standardization (or customization) here is a problem when it comes to reproducability and consistency for local models; one backend's set of sampler settings has a totally different effect on another backend (such as text-generation-webui's HF loaders, which assume the transformers order of temp first.)

**Comment by KerfuffleV2** [comment]
Independent projects can make their own choices. They aren't really "backends", if anything, GGML or llama.cpp as a lib is the backend. 

Better customizability is obviously great though.
